frame/dvm/src/lib.rs

```diff
@@ -298,139 +297,52 @@ pub mod pallet {
-			receipts.push(receipt.clone());
-			Self::logs_bloom(receipt.logs.clone(), &mut logs_bloom);
-		}
-
-		let ommers = Vec::<ethereum::Header>::new();
-		let partial_header = ethereum::PartialHeader {
-			parent_hash: Self::current_block_hash().unwrap_or_default(),
-			beneficiary: darwinia_evm::Pallet::<T>::find_author(),
-			// TODO: figure out if there's better way to get a sort-of-valid state root.
-			state_root: H256::default(),
-			receipts_root: H256::from_slice(
-				Keccak256::digest(&rlp::encode_list(&receipts)[..]).as_slice(),
-			), // TODO: check receipts hash.
-			logs_bloom,
-			difficulty: U256::zero(),
-			number: block_number,
-			gas_limit: T::BlockGasLimit::get(),
-			gas_used: receipts
-				.clone()
-				.into_iter()
-				.fold(U256::zero(), |acc, r| acc + r.used_gas),
-			timestamp: UniqueSaturatedInto::<u64>::unique_saturated_into(
-				<pallet_timestamp::Pallet<T>>::get(),
-			),
-			extra_data: Vec::new(),
-			mix_hash: H256::default(),
-			nonce: H64::default(),
-		};
-		let mut block = ethereum::Block::new(partial_header, transactions.clone(), ommers);
-		block.header.state_root = T::StateRoot::get();
-
-		CurrentBlock::<T>::put(block.clone());
-		CurrentReceipts::<T>::put(receipts.clone());
-		CurrentTransactionStatuses::<T>::put(statuses.clone());
-		BlockHash::<T>::insert(block_number, block.header.hash());
-
-		if post_log {
-			let digest = DigestItem::<T::Hash>::Consensus(
-				FRONTIER_ENGINE_ID,
-				PostLog::Hashes(dp_consensus::Hashes::from_block(block)).encode(),
-			);
-			<frame_system::Pallet<T>>::deposit_log(digest.into());
-		}
-	}
-
-	fn logs_bloom(logs: Vec<Log>, bloom: &mut Bloom) {
-		for log in logs {
-			bloom.accrue(BloomInput::Raw(&log.address[..]));
-			for topic in log.topics {
-				bloom.accrue(BloomInput::Raw(&topic[..]));
-			}
-		}
-	}
-
+	/// Execute transaction from pallet(internal transaction)
+	/// NOTE: The difference between the rpc transaction and the internal transaction is that
+	/// The internal transactions will catch and throw evm error comes from runner to caller.
 	pub fn internal_transact(target: H160, input: Vec<u8>) -> DispatchResultWithPostInfo {
 		ensure!(
 			dp_consensus::find_pre_log(&<frame_system::Pallet<T>>::digest()).is_err(),
 			Error::<T>::PreLogExists,
 		);
+
 		let nonce =
 			<T as darwinia_evm::Config>::RingAccountBasic::account_basic(&INTERNAL_CALLER).nonce;
-		let transaction = DVMTransaction::new(nonce, target, input);
-
-		Self::raw_transact(transaction)
+		let transaction = DVMTransaction::new_internal_transaction(nonce, target, input);
+		Self::raw_transact(transaction).map(|(reason, used_gas)| match reason {
+			// Only when exit_reason is successful, return Ok(...)
+			ExitReason::Succeed(_) => Ok(PostDispatchInfo {
+				actual_weight: Some(T::GasWeightMapping::gas_to_weight(
+					used_gas.unique_saturated_into(),
+				)),
+				pays_fee: Pays::No,
+			}),
+			_ => {
+				log::error!("Executing internal transaction error happened");
```

Comment:
Still need this?

Reply:
Did you mean this log line or this match branch `_`?

Reply:
The log.

The error can be caught outside now.

Reply:
Just keep it.